src/main/scala/is/hail/utils/BinarySearch.scala

```diff
@@ -0,0 +1,51 @@
+package is.hail.utils
+
+object BinarySearch {
+  // return smallest elem such that key <= elem
+  def binarySearch(length: Int,
+    // key.compare(elem)
+    compare: (Int) => Int): Int = {
+    assert(length > 0)
+
+    var low = 0
+    var high = length - 1
+    while (low < high) {
+      val mid = (low + high) / 2
+      assert(mid >= low && mid < high)
+
+      // key <= elem
+      if (compare(mid) <= 0) {
+        high = mid
+      } else {
+        low = mid + 1
+      }
+    }
+    assert(low == high)
+    assert(low >= 0 && low < length)
+
+    // key <= low
+    assert(compare(low) <= 0 || low == length - 1)
+    // low == 0 || (low - 1) > key
+    assert(low == 0
+      || compare(low - 1) > 0)
+
+    low
+  }
+  
+  // a is increasing and may contain duplicate values. Returns:
+  //      -1 iff a is empty or key < a(0)
+  //       j iff a(j) <= key < a(j + 1)
+  //   len-1 iff a(len - 1) < key
+  def binarySearchInterval(a: Array[Long], key: Long): Int = {
```

Comment:
Consider:
```scala
def foo(arr: Array[Long], key: Long): Int =
  binarySearch(arr.length, i => (key - arr(i)).toInt)
```

This is very similar to `binarySearchInterval`. The only difference is that you want `key < a(0)` to be `-1`.

Why do you want to return `-1`? In the use-case of `getPartition`, shouldn't the element _always_ exist in the array?

Reply:
This works correctly on the use case of getPartition, and happens to have nice properties even for keys outside of the array values (-1 and len-1 are natural values to return when the element is < or >= partitionStarts).

In any case, `key - arr(i)` need not be an Int, and here's the comparison for keys in range `[0, 0, 0, 4, 5, 5, 8, 10, 10]` showing very different behavior:
```
key, BinarySearch.binarySearchInterval(a, i), foo(a, i)
(0,2,0)
(1,2,3)
(2,2,3)
(3,2,3)
(4,3,3)
(5,5,4)
(6,5,6)
(7,5,6)
(8,6,6)
(9,6,7)
```
My function is the right one for partitionStarts on RowMatrix, and I think when OrderedRVD2 is in there will be both starts and ends available for each partition and opportunity to move to partitionStarts instead of rangeBounds there as well (which I think is more natural). Happy to discuss more in person.